Allow inspect to work even if tag list blocked

estesp opened the pull request:
Some registries may choose to block the "list all tags" endpoint for
performance or other reasons. In this case we should still allow an
inspect which will not include the "tag list" in the output.

Signed-off-by: Phil Estes <estesp@gmail.com>

runcom commented on cmd/skopeo/inspect.go:
```diff
@@ -97,7 +99,13 @@ var inspectCmd = cli.Command{
-				return fmt.Errorf("Error determining repository tags: %v", err)
+				// some registries may decide to block the "list all tags" endpoint
+				// gracefully allow the inspect to continue in this case
+				if !strings.Contains(err.Error(), "401") {
```
should we check this in containers/image and just returns `[]` there? @mtrmac wdyt?

mtrmac replied:
Not silently, there is a difference between “definitely an untagged image” and “server refuses to tell us”.

But returning a typed error from `GetRepositoryTags` so that this wouldn’t have to do a `strings.Contains` would be nice.

mtrmac replied:
[BTW it would be nice to have a documented example of “some registries”, perhaps naming a product or a product version or a config option, either in the code comment or at least mentioned in this PR.

That would allow other users to find this PR and to realize that they need to upgrade.

estesp replied:
@mtrmac yes.. I should have explained more. The IBM Bluemix container registry is a v2 registry that, because of original issues with a specific backend and performance blocked the "get all tags".  This restriction will probably be removed, but for now, I decided to fix the few places I knew where tools were failing when targeting a Bluemix repository.

runcom commented:
Thanks for this @estesp :heart:

mtrmac commented:
LGTM I guess, the possible improvements mentioned above are not quite blockers from my POV; will defer to @runcom on whether to merge as is.

@estesp FWIW the tests are failing on golint.

estesp commented:
Updated to fix the `go lint` issue; also added a specific comment about the Bluemix registry restriction (which will probably go away).

I also agree a typed error would make this side cleaner than matching on the 401 error return in the string. I'm fine if there is a desire to clean up that side first and then add this. Either way is fine with me.

mtrmac commented:
Thanks! I can live with the string matching here, but I’ll defer to @runcom .

runcom commented:
Well yeah, it's fine with me even with this for now, I'll work to port this on c/image

LGTM

mtrmac commented:
@estesp If this is still rejected by the IBM Bluemix registry, do you have sample registry response that is returned by an attempt to list tags?

It would be useful to have a permanent test case in https://github.com/containers/image/blob/d2420cbbf4d8c44e827bbeb7ce24abec211fba65/docker/errors_test.go#L18 , to make sure we don’t regress this.

mtrmac commented:
For the record:
- AFAICT, “IBM Bluemix” has become “IBM Cloud”, and the “Bluemix” registry is now (somehow related to?) `icr.io`; e.g. https://cloud.ibm.com/docs/Registry?topic=Registry-registry_overview lists `bluemix.net` and `icr.io` host names.
- Randomly looking for a public image hosted on that registry, at least `bin/skopeo list-tags docker://icr.io/codeengine/firstjob` now succeeds.

So I’m assuming that at least the current cloud deployment now allows listing tags, and does not need special handling. (Is that true for all currently deployed versions?)
